diff -Nur alsa-utils-1.0.24.2_orig//alsamixer/volume_mapping.c alsa-utils-1.0.24.2/alsamixer/volume_mapping.c
--- alsa-utils-1.0.24.2_orig//alsamixer/volume_mapping.c	2012-03-07 13:32:42.938486109 +0100
+++ alsa-utils-1.0.24.2/alsamixer/volume_mapping.c	2012-03-07 13:34:19.457272705 +0100
@@ -39,6 +39,12 @@
 
 #define MAX_LINEAR_DB_SCALE	24
 
+static inline  double _exp10(double x)
+{
+	return exp(x * log(10));
+}
+
+
 static inline bool use_linear_dB_scale(long dBmin, long dBmax)
 {
 	return dBmax - dBmin <= MAX_LINEAR_DB_SCALE * 100;
@@ -109,9 +115,9 @@
 	if (use_linear_dB_scale(min, max))
 		return (value - min) / (double)(max - min);
 
-	normalized = exp10((value - max) / 6000.0);
+	normalized = _exp10((value - max) / 6000.0);
 	if (min != SND_CTL_TLV_DB_GAIN_MUTE) {
-		min_norm = exp10((min - max) / 6000.0);
+		min_norm = _exp10((min - max) / 6000.0);
 		normalized = (normalized - min_norm) / (1 - min_norm);
 	}
 
@@ -144,7 +150,7 @@
 	}
 
 	if (min != SND_CTL_TLV_DB_GAIN_MUTE) {
-		min_norm = exp10((min - max) / 6000.0);
+		min_norm = _exp10((min - max) / 6000.0);
 		volume = volume * (1 - min_norm) + min_norm;
 	}
 	value = lrint_dir(6000.0 * log10(volume), dir) + max;
