--- ffmpeg_tiny_org/configure	2011-08-03 18:48:48.000000000 +0200
+++ ffmpeg_tiny/configure	2011-08-03 18:50:37.000000000 +0200
@@ -2620,7 +2620,7 @@
     enabled armv6t2 && check_asm armv6t2 '"movt r0, #0"'
     enabled armvfp  && check_asm armvfp  '"fadds s0, s0, s0"'
     enabled iwmmxt  && check_asm iwmmxt  '"wunpckelub wr6, wr4"'
-    enabled neon    && check_asm neon    '"vadd.i16 q0, q0, q0"'
+    enabled neon    && check_asm neon    '"vadd.i16 q0, q0, q0"' -mfpu=neon
     enabled vfpv3   && check_asm vfpv3   '"vmov.f32 s0, #1.0"'
 
     enabled_all armv6t2 shared !pic && enable_pic
